### PR TITLE
Refactor Blueprint to C++

### LINKED ISSUE #5: Migrate Preliminary Blueprint to C++
### Move spaceship behavior to code 

* [x] Convert all axes handlers to C++
* [x] Handle particle systems for thrust effects in C++

### It works

* [x] Everything functions the same as before!...

### PR DESCRIPTION
**Pull request by CheerfulSatchel**
Closes #5.

# Rewrite preliminary Blueprint code to C++ 

Converted the initial Blueprint code to C++. The main motivation was to better familiarize ourselves with Unreal's C++ libraries and to inflict additional pain on ourselves (LOL).

Some of the Blueprint code was getting fairly cluttered, so now we can get C++ clutter instead 😃👍 

Spaceship flies like a lander and explodes when an arbitrary normal impulse magnitude of force hits the ship.

# How was it tested?

Local testing, lots of trial and error.

# Additional Notes

<sub>This really wasn't necessary, but I was bored 😂</sub>

### CODE REVIEW

File: Source/LanderLads/SpaceshipPawn.cpp [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+	this->AddControllerPitchInput(
+		this->GetWorld()->GetDeltaSeconds() * -45 * ClampedAxisValue);
+}
+
+const float ASpaceshipPawn::ClampAxisValue(float AxisValue)
+{
+	return FMath::Clamp(AxisValue, -1.0f, 1.0f);
+}
+
+USpaceshipStaticMeshComponent* ASpaceshipPawn::GetSpaceshipStaticMeshComponent()
```
Comment (Lancewiu) [inline Source/LanderLads/SpaceshipPawn.cpp]: I'm actually pretty sure this results in longer lines of code.

File: Source/LanderLads/SpaceshipPawn.cpp [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+}
+
+void ASpaceshipPawn::MoveCameraLookVertical(float AxisValue)
+{
+	float ClampedAxisValue = ClampAxisValue(AxisValue);
+	this->AddControllerPitchInput(
+		this->GetWorld()->GetDeltaSeconds() * -45 * ClampedAxisValue);
+}
+
+const float ASpaceshipPawn::ClampAxisValue(float AxisValue)
```
Comment (Lancewiu) [inline Source/LanderLads/SpaceshipPawn.cpp]: Dangerous as it hides magic values inside a one-liner.  If the clamp must be [-1.0f, 1.0f] then justify it in the context.

File: Source/LanderLads/SpaceshipPawn.cpp [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+}
+
+// Called to bind functionality to input
+void ASpaceshipPawn::SetupPlayerInputComponent(UInputComponent* PlayerInputComponent)
+{
+	Super::SetupPlayerInputComponent(PlayerInputComponent);
+
+	check(PlayerInputComponent);
+
+	PlayerInputComponent->BindAction("MoveUp", EInputEvent::IE_Pressed, this, &ASpaceshipPawn::MoveUpButtonPressed);
```
Comment (Lancewiu) [inline Source/LanderLads/SpaceshipPawn.cpp]: Why bind action if the input result works well with axis?

File: Source/LanderLads/SpaceshipPawn.cpp [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+	UPrimitiveComponent* OtherComp,
+	FVector NormalImpulse,
+	const FHitResult& Hit)
+{
+	if (HitComp == this->GetSpaceshipStaticMeshComponent()) {
+		constexpr float ForceThreshold = 300000;
+		if (NormalImpulse.X > ForceThreshold ||
+			NormalImpulse.Y > ForceThreshold ||
+			NormalImpulse.Z > ForceThreshold) {
+			this->ExplodeShip();
```
Comment (Lancewiu) [inline Source/LanderLads/SpaceshipPawn.cpp]: If you only use this once, just put the function in here so we don't pollute the header file.

File: Source/LanderLads/SpaceshipPawn.cpp [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+void ASpaceshipPawn::MoveCameraLookVertical(float AxisValue)
+{
+	float ClampedAxisValue = ClampAxisValue(AxisValue);
+	this->AddControllerPitchInput(
+		this->GetWorld()->GetDeltaSeconds() * -45 * ClampedAxisValue);
+}
+
+const float ASpaceshipPawn::ClampAxisValue(float AxisValue)
+{
+	return FMath::Clamp(AxisValue, -1.0f, 1.0f);
```
Comment (Lancewiu) [inline Source/LanderLads/SpaceshipPawn.cpp]: This function is dangerous as it hides magic values inside a one-liner.  If the clamp must be [-1.0f, 1.0f] then justify it in the context.

File: Source/LanderLads/SpaceshipPawn.cpp [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+}
+
+USpaceshipStaticMeshComponent* ASpaceshipPawn::GetSpaceshipStaticMeshComponent()
+{
+	return this->SpaceshipStaticMeshComponent;
```
Comment (Lancewiu) [inline Source/LanderLads/SpaceshipPawn.cpp]: This getter function results in longer lines of code.
Furthermore, getters should be reserved for public APIs and not private ones.  It is generally assumed that the class knows how to handle itself.

File: Source/LanderLads/SpaceshipPawn.cpp [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+	Super::SetupPlayerInputComponent(PlayerInputComponent);
+
+	check(PlayerInputComponent);
+
+	PlayerInputComponent->BindAction("MoveUp", EInputEvent::IE_Pressed, this, &ASpaceshipPawn::MoveUpButtonPressed);
+	PlayerInputComponent->BindAction("MoveDown", EInputEvent::IE_Pressed, this, &ASpaceshipPawn::MoveDownButtonPressed);
+	PlayerInputComponent->BindAction("MoveRight", EInputEvent::IE_Pressed, this, &ASpaceshipPawn::MoveRightButtonPressed);
+	PlayerInputComponent->BindAction("MoveLeft", EInputEvent::IE_Pressed, this, &ASpaceshipPawn::MoveLeftButtonPressed);
+	PlayerInputComponent->BindAction("MoveForward", EInputEvent::IE_Pressed, this, &ASpaceshipPawn::MoveForwardButtonPressed);
+	PlayerInputComponent->BindAction("MoveBackward", EInputEvent::IE_Pressed, this, &ASpaceshipPawn::MoveBackwardButtonPressed);
```
Comment (Lancewiu) [inline Source/LanderLads/SpaceshipPawn.cpp:56]: Why bind action if the input result works well with axis?

File: Source/LanderLads/SpaceshipPawn.cpp (hunk 5 above)
Comment (CheerfulSatchel) [reply Source/LanderLads/SpaceshipPawn.cpp]: Great point! 👍

File: Source/LanderLads/SpaceshipPawn.cpp (hunk 6 above)
Comment (CheerfulSatchel) [reply Source/LanderLads/SpaceshipPawn.cpp]: Nice tip !!

File: Source/LanderLads/SpaceshipPawn.cpp (hunk 7 above)
Comment (CheerfulSatchel) [reply Source/LanderLads/SpaceshipPawn.cpp:56]: Action binding is primarily for toggling the thrusters 😃

**Review by Lancewiu — APPROVED** [review]

**Pull request merged**